arviz/data/io_pyro.py

```diff
@@ -1,119 +1,123 @@
+        for i, k in enumerate(sorted(divergences)):
+            diverging[i, divergences[k]] = 1
+        data = {"diverging": diverging}
+        return dict_to_dataset(data, library=self.pyro, coords=self.coords, dims=self.dims)
 
-        try:  # Try pyro>=0.3 release syntax
-            data = {
-                name: utils.expand_dims(samples.enumerate_support().squeeze())
-                for name, samples in self.posterior.marginal(
-                    sites=self.observed_vars
-                ).empirical.items()
-            }
-        except AttributeError:  # Use pyro<0.3 release syntax
-            data = {}
-            for var_name in self.observed_vars:
-                # pylint: disable=no-member
-                samples = EmpiricalMarginal(
-                    self.posterior, sites=var_name
-                ).get_samples_and_weights()[0]
-                samples = samples.numpy().squeeze()
-                data[var_name] = utils.expand_dims(samples)
+    @requires("posterior_predictive")
+    def posterior_predictive_to_xarray(self):
+        """Convert posterior_predictive samples to xarray."""
+        data = {}
+        for k, ary in self.posterior_predictive.items():
+            ary = ary.detach().cpu().numpy()
+            shape = ary.shape
+            if shape[0] == self.nchains and shape[1] == self.ndraws:
+                data[k] = ary
+            elif shape[0] == self.nchains * self.ndraws:
+                data[k] = ary.reshape((self.nchains, self.ndraws, *shape[1:]))
+            else:
+                data[k] = utils.expand_dims(ary)
+                _log.warning(
+                    "posterior predictive shape not compatible with number of chains and draws. "
+                    "This can mean that some draws or even whole chains are not represented."
+                )
         return dict_to_dataset(data, library=self.pyro, coords=self.coords, dims=self.dims)
 
+    @requires("prior")
+    def prior_to_xarray(self):
+        """Convert prior samples to xarray."""
+        return dict_to_dataset(
+            {k: utils.expand_dims(v.detach().cpu().numpy()) for k, v in self.prior.items()},
+            library=self.pyro,
+            coords=self.coords,
+            dims=self.dims,
+        )
+
     def to_inference_data(self):
         """Convert all available data to an InferenceData object."""
         return InferenceData(
             **{
                 "posterior": self.posterior_to_xarray(),
-                "observed_data": self.observed_data_to_xarray(),
+                "sample_stats": self.sample_stats_to_xarray(),
+                "posterior_predictive": self.posterior_predictive_to_xarray(),
+                "prior": self.prior_to_xarray(),
             }
         )
 
 
-def from_pyro(posterior=None, *, coords=None, dims=None):
-    """Convert pyro data into an InferenceData object.
+def from_pyro(posterior=None, *, prior=None, posterior_predictive=None, coords=None, dims=None):
```

Comment:
Thanks for explaining! I understand how arviz works better now. Will address this...

Reply:
I decide to follow io_pymc and set nchains, ndraws to 0 if posterior=None. Is it reasonable to do so?

Reply:
It is, these values are only stored in the converter to check posterior predictive shape in io_pymc. I don't know if maybe in pyro it would make sense to check/reshape prior samples too?

Reply:
Thanks! Then I will stick with this pattern.